compendium/DeclarativeServices/test/TestUpdateConfiguration.cpp

```diff
@@ -0,0 +1,517 @@
+} // end of testUpdateConfigWithoutModifiedMethodImmediate
+
+/**
+  *  Test component instance won't be constructed or activated until requested.
+  */
+TEST_F(tServiceComponent,
+       testUpdateConfig_WithoutModifiedMethodDelayed) // DS_CAI_FTC_4
+{
+  // Start the test bundle containing the component name.
+  std::string componentName = "sample::ServiceComponentCA04";
+  cppmicroservices::Bundle testBundle = StartTestBundle("TestBundleDSCA04");
+
+  // Use DS runtime service to validate the component state
+  scr::dto::ComponentDescriptionDTO compDescDTO;
+  auto compConfigs =
+    GetComponentConfigs(testBundle, componentName, compDescDTO);
+  EXPECT_EQ(compConfigs.size(), 1ul) << "One default config expected";
+  EXPECT_EQ(compConfigs.at(0).state,
+            scr::dto::ComponentState::UNSATISFIED_REFERENCE)
+    << "The state should be UNSATISFIED_REFERENCE.";
+
+  // Get a service reference to ConfigAdmin to create the component instance.
+  auto configAdminService =
+    GetInstance<cppmicroservices::service::cm::ConfigurationAdmin>();
+  ASSERT_TRUE(configAdminService) << "GetService failed for ConfigurationAdmin";
+
+  // Create configuration object and update property.
+  auto configuration = configAdminService->GetConfiguration(componentName);
+  auto configInstance = configuration->GetPid();
+
+  cppmicroservices::AnyMap props(
+    cppmicroservices::AnyMap::UNORDERED_MAP_CASEINSENSITIVE_KEYS);
+  const std::string instanceId{ "instance1" };
+  props["uniqueProp"] = instanceId;
+  auto fut = configuration->Update(props);
+  fut.get();
+  // GetService to make component active.
+  auto instance = GetInstance<test::CAInterface>();
+  ASSERT_TRUE(instance) << "GetService failed for CAInterface.";
+
+  // Confirm configuration object presented and check component state.
+  compConfigs = GetComponentConfigs(testBundle, componentName, compDescDTO);
+  EXPECT_EQ(compConfigs.size(), 1ul) << "One default config expected.";
+  EXPECT_EQ(compConfigs.at(0).state, scr::dto::ComponentState::ACTIVE)
+    << "Component instance state should be ACTIVE.";
+
+  // Confirm component instance was created with the correct properties.
+  auto instanceProps = instance->GetProperties();
+  auto uniqueProp = instanceProps.find("uniqueProp");
+
+  ASSERT_TRUE(uniqueProp != instanceProps.end())
+    << "uniqueProp not found in constructed instance.";
+  EXPECT_EQ(uniqueProp->second, instanceId);
+
+  // Update property
+  const std::string instanceId2{ "instance2" };
+  props["uniqueProp"] = instanceId2;
+  fut = configuration->Update(props);
+  fut.get();
+  compConfigs = GetComponentConfigs(testBundle, componentName, compDescDTO);
+  EXPECT_EQ(compConfigs.at(0).state, scr::dto::ComponentState::SATISFIED)
+    << "Component state should be SATISFIED";
+
+  // This will cause DS to construct the instance with the updated properties.
+  auto newInstance = GetInstance<test::CAInterface>();
+  ASSERT_TRUE(newInstance) << "GetService failed for CAInterface";
```

Comment:
if testing for activation, I would think that the component should be checked that its in the scr::dto::ComponentState::ACTIVE state.

Reply:
Lines 295 - 301 use the GetProperties method of the component instance to verify the property values. It wouldn't be able to execute this method if the component wasn't ACTIVE. We thought this was enough of a confirmation that the component is active but we could add a state check to confirm that the state is scr::dto::ComponentState::ACTIVE if you think we should.

Reply:
I think it's better not to rely on an indirect behavior (i.e. the behavior of `GetProperties`) to verify a test point and instead query the state directly.

Reply:
Done